[quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45881 [quant] Refactor qembeddingbag to remove duplicate code
* #45865 [quant] Support for 4-bit quantized EmbeddingBag module
* #45752 [quant] Support 4-bit embedding_bag operators using the dtype quint4x2
* **#45751 [quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2**

Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120997](https://our.internmc.facebook.com/intern/diff/D24120997)

supriyar committed:
[quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2
Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -14,21 +14,35 @@ torch::class_<EmbeddingPackedParamsBase> register_embedding_params();
+  uint8_t* weight_data;
+  int bit_width, scale_bias_bytes;
+  if (qweight.scalar_type() == c10::kQUInt8) {
+    weight_data =
+        reinterpret_cast<uint8_t*>(weight_contig.data_ptr<c10::quint8>());
```
I remember data_ptr does not do type check, so maybe you can just do weight_contig.data_ptr<uint8_t*>

supriyar replied:
I don't think it is applicable for qint types. It throws the error "expected scalar type Byte but found QUInt8"

jerryzh168 replied:
oh sorry, I meant this: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/TensorBody.h#L354
so:
```
static_cast<uint8_t*>(weight_contig.data_ptr())
```

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_unpack.cpp:
```diff
@@ -9,38 +9,71 @@ torch::class_<EmbeddingPackedParamsBase> register_embedding_params();
+          zero_points.toType(c10::kFloat),
+          0, // The output channel axis is 0
+          device(c10::kCPU).dtype(c10::kQUInt4x2));
+      output_data =
```
same here

supriyar committed:
Update on "[quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -61,28 +76,46 @@ c10::intrusive_ptr<EmbeddingPackedParamsBase> PackedEmbeddingBagWeight::prepack(
+    // in a byte.
+    embedding_cols =
+        (embedding_cols + num_elem_per_byte - 1) / num_elem_per_byte;
+    at::parallel_for(
+        0, embedding_rows, 1, [&](int32_t start_idx, int32_t end_idx) {
+          for (int64_t row = start_idx; row < end_idx; ++row) {
+            const uint8_t* input_row = weight_data + row * embedding_cols;
+            std::uint8_t* output_row = output_data + row * output_columns;
+            at::Half* output_row_scale_bias =
+                reinterpret_cast<at::Half*>(output_row + embedding_cols);
+            output_row_scale_bias[0] = weight_scales[row];
+            output_row_scale_bias[1] = weight_bias[row];
+            for (int64_t col = 0; col < embedding_cols; ++col) {
+              // The weight values have already been packed, so here we just
+              // store it in the output tensor.
+              output_row[col] = input_row[col];
+            }
+          }
+        });
```
optional: seems like this could be reused with the above section if the `at::Half` is templatized?  Def optional though since LOC is low.

supriyar replied:
Yes, I think there would be similar LOC added in templatizing it so planning to skip it for now.

supriyar committed:
Update on "[quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120997](https://our.internmc.facebook.com/intern/diff/D24120997)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120997](https://our.internmc.facebook.com/intern/diff/D24120997)

[ghstack-poisoned]
